### DISCUSSION TITLE
Adding Matrix Stack support for Transform?D

### DISCUSSION THREAD

**Discussion by dayo05** [godotengine/godot-proposals#6673] [GUI]
In GodotSharp, there are Transform?D class which used for custom drawing, that code looks similar as OpenGL, DX.
I suggest to add matrix stack feature which legacy OpenGL and some extra libraries for modern OpenGL, DirectXMath supports.

This feature will useful for who wants to render a lots of things or complicated things manually in `_Draw` function.

I suggest to add Transform?D.Push() and Transform?D.Pop() method or Adding external Matrix Stack class for Transform.

**Comment by theraot** [comment]
`CanvasItem` could use `draw_pop_transform_matrix` and `draw_push_transform_matrix`. Note that `CanvasItem` already has [`draw_set_transform_matrix`](https://docs.godotengine.org/en/latest/classes/class_canvasitem.html#class-canvasitem-method-draw-set-transform-matrix). Placing this in `CanvasItem` could allow optimization for rendering. I believe that should be sufficient for your use case and avoid the issues I mention below.

---

The next best option - in my opinion - is a dedicated type. I don't think that could be optimized because you could use the dedicated type in any context, but it would give you the ease of use. Plus it would be natural for it to return the identity transform when empty.

Baring that, `Array` could get `peek_front` and `peek_back`. Which would help a bit with the ease compared to using an `Array` for this today, but it would not be as good as having a dedicated type. And of course, it could not be optimized either.

*I think making a packed array type would have other impacts because they are also cases for Variant. I also think that adding this to every transform would be detrimental for other places where transforms are used.*
